src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,438 @@
+            {
+                PatternSyntax pattern => pattern,
+                ExpressionSyntax constant => CreatePattern(receiver, constant, flipped),
+                var v => throw ExceptionUtilities.UnexpectedValue(v),
+            };
+        }
+
+        private static (PatternSyntax ContainingPattern, ImmutableArray<IdentifierNameSyntax> NamesOpt)? TryFindVariableDesignation(
+            PatternSyntax leftPattern,
+            ExpressionSyntax rightReceiver,
+            SemanticModel model)
+        {
+            if (GetInnermostReceiver(rightReceiver, model) is not IdentifierNameSyntax identifierName)
+                return null;
+
+            var designation = leftPattern.DescendantNodes()
+                .OfType<SingleVariableDesignationSyntax>()
+                .Where(d => AreEquivalent(d.Identifier, identifierName.Identifier))
+                .FirstOrDefault();
+
+            // For simplicity, we only support replacement when the designation is contained in one of the following patterns.
+            // This excludes a parenthesized variable designation, for example, which would require rewriting the whole thing.
+            var parent = designation.Parent;
+            if (!parent.IsKind(SyntaxKind.VarPattern, SyntaxKind.DeclarationPattern, SyntaxKind.RecursivePattern))
+                return null;
+
+            // Since we're looking for a variable designation, we permit a replacement with no subpatterns.
+            // For instance, `e is C v && v is { p: 1 }` would be converted to `e is C { p: 1 } v`
+            _ = TryGetParentNames(identifierName, out var names);
+            return ((PatternSyntax)parent, names);
+        }
+
+        private static PatternSyntax CreatePattern(ExpressionSyntax receiver, ExpressionSyntax target, bool flipped)
+        {
+            return receiver.Parent switch
+            {
+                BinaryExpressionSyntax(EqualsExpression) => ConstantPattern(target),
+                BinaryExpressionSyntax(NotEqualsExpression) => UnaryPattern(ConstantPattern(target)),
+                BinaryExpressionSyntax(GreaterThanExpression or
+                                       GreaterThanOrEqualExpression or
+                                       LessThanOrEqualExpression or
+                                       LessThanExpression) e
+                    => RelationalPattern(flipped ? Flip(e.OperatorToken) : e.OperatorToken, target),
+                var v => throw ExceptionUtilities.UnexpectedValue(v),
+            };
+
+            static SyntaxToken Flip(SyntaxToken token)
+            {
+                var kind = token.Kind() switch
+                {
+                    LessThanToken => GreaterThanToken,
+                    LessThanEqualsToken => GreaterThanEqualsToken,
+                    GreaterThanEqualsToken => LessThanEqualsToken,
+                    GreaterThanToken => LessThanToken,
+                    var v => throw ExceptionUtilities.UnexpectedValue(v)
```

Comment:
```suggestion
                    _ => throw ExceptionUtilities.UnexpectedValue(token)
```

Reply:
I'd prefer to use var patterns for this universally to ensure we pass the actual input variable.